src/TurbulenceConvection/EDMF_functions.jl

```diff
@@ -0,0 +1,1352 @@
+            qi_tendency_precip_formation = aux_up_i.qi_tendency_precip_formation
+
+            ρaq_liq = prog_up[i].ρaq_liq
+            ρaq_ice = prog_up[i].ρaq_ice
+
+            tends_ρaq_liq = tendencies_up[i].ρaq_liq
+            tends_ρaq_ice = tendencies_up[i].ρaq_ice
+
+            @. tends_ρaq_liq =
+                -∇c(wvec(LBF(Ic(w_up) * ρaq_liq))) +
+                (ρarea * Ic(w_up) * entr_turb_dyn * q_liq_en) -
+                (ρaq_liq * Ic(w_up) * detr_turb_dyn) +
+                (ρ_c * (ql_tendency_precip_formation + ql_tendency_noneq))
+
+            @. tends_ρaq_ice =
+                -∇c(wvec(LBF(Ic(w_up) * ρaq_ice))) +
+                (ρarea * Ic(w_up) * entr_turb_dyn * q_ice_en) -
+                (ρaq_ice * Ic(w_up) * detr_turb_dyn) +
+                (ρ_c * (qi_tendency_precip_formation + qi_tendency_noneq))
+
+            tends_ρaq_liq[kc_surf] = 0
+            tends_ρaq_ice[kc_surf] = 0
+        end
+
+        # prognostic entr/detr
+        if edmf.entr_closure isa PrognosticNoisyRelaxationProcess
+            c_gen_stoch = edmf.entr_closure.c_gen_stoch
+            mean_entr = aux_up[i].ε_nondim
+            mean_detr = aux_up[i].δ_nondim
+            ε_λ = c_gen_stoch[3]
+            δ_λ = c_gen_stoch[4]
+            tends_ε_nondim = tendencies_up[i].ε_nondim
+            tends_δ_nondim = tendencies_up[i].δ_nondim
+            ε_nondim = prog_up[i].ε_nondim
+            δ_nondim = prog_up[i].δ_nondim
+            @. tends_ε_nondim = ε_λ * (mean_entr - ε_nondim)
+            @. tends_δ_nondim = δ_λ * (mean_detr - δ_nondim)
+        end
+
+        tends_ρarea[kc_surf] = 0
+        tends_ρaθ_liq_ice[kc_surf] = 0
+        tends_ρaq_tot[kc_surf] = 0
+    end
+
+    # Solve for updraft velocity
+
+    # We know that, since W = 0 at z = 0, BCs for entr, detr,
+    # and buoyancy should not matter in the end
+    zero_bcs = (; bottom = CCO.SetValue(FT(0)), top = CCO.SetValue(FT(0)))
+    I0f = CCO.InterpolateC2F(; zero_bcs...)
+    adv_bcs =
+        (; bottom = CCO.SetValue(wvec(FT(0))), top = CCO.SetValue(wvec(FT(0))))
+    LBC = CCO.LeftBiasedF2C(; bottom = CCO.SetValue(FT(0)))
+    ∇f = CCO.DivergenceC2F(; adv_bcs...)
+
+    @inbounds for i in 1:N_up
+        a_up_bcs = a_up_boundary_conditions(surf, edmf, i)
+        Iaf = CCO.InterpolateC2F(; a_up_bcs...)
+        ρaw = prog_up_f[i].ρaw
+        tends_ρaw = tendencies_up_f[i].ρaw
+        nh_pressure = aux_up_f[i].nh_pressure
+        a_up = aux_up[i].area
+        w_up = aux_up_f[i].w
+        w_en = aux_en_f.w
+        entr_w = aux_up[i].entr_turb_dyn
+        detr_w = aux_up[i].detr_turb_dyn
+        buoy = aux_up[i].buoy
+
+        @. tends_ρaw = -(∇f(wvec(LBC(ρaw * w_up))))
+        @. tends_ρaw +=
+            (ρaw * (I0f(entr_w) * w_en - I0f(detr_w) * w_up)) +
+            (ρ_f * Iaf(a_up) * I0f(buoy)) +
+            nh_pressure
+        tends_ρaw[kf_surf] = 0
+    end
+
+    return nothing
+end
+
+function filter_updraft_vars(
```

Comment:
Can we call this `crop_updraft_vars` because it is clipping/cropping, rather than filtering (that term suggests conservation properties that we do not have here).